proposals/p2188.md

```diff
@@ -0,0 +1,664 @@
+}
+```
+
+If some fields should be ignored when matching, a trailing `, _` can be added to
+specify this:
+
+```
+match ({.a = 1, .b = 2}) {
+  case {.a = 1, _} => { return 1; }
+  case {b: i32, _} => { return b; }
+}
+```
+
+This is valid even if all fields are actually named in the pattern.
+
+### Alternative patterns
+
+An alternative pattern is used to match one alternative of a choice type.
+
+-   _proper-pattern_ ::= _callee-expression_ _tuple-pattern_
+-   _proper-pattern_ ::= _designator_ _tuple-pattern_?
+
+Here, _callee-expression_ is syntactically an expression that is valid as the
+callee in a function call expression, and an alternative pattern is
+syntactically a function call expression whose argument list contains at least
+one _proper-pattern_.
+
+If a _callee-expression_ is provided, it is required to name a choice type
+alternative that has a parameter list, and the scrutinee is implicitly converted
+to that choice type. Otherwise, the scrutinee is required to be of some choice
+type, and the designator is looked up in that type and is required to name an
+alternative with a parameter list if and only if a _tuple-pattern_ is specified.
+
+The pattern matches if the active alternative in the scrutinee is the specified
+alternative, and the arguments of the alternative match the given tuple pattern
+(if any).
+
+```
+choice Optional(T:! Type) {
+  None,
+  Some(T)
+}
+
+match (Optional(i32).None) {
+  // ✅ `.None` resolved to `Optional(i32).None`.
+  case .None => {}
+  // ✅ `.Some` resolved to `Optional(i32).Some`.
+  case .Some(n: i32) => { Print("{0}", n); }
+  // ❌ Error, no such alternative exists.
+  case .Other => {}
+}
+
+class X {
+  external impl as ImplicitAs(Optional(i32));
+}
+
+match ({} as X) {
+  // ✅ OK, but expression pattern.
+  case Optional(i32).None => {}
+  // ✅ OK, implicitly converts to `Optional(i32)`.
+  case Optional(i32).Some(n: i32) => { Print("{0}", n); }
+}
+```
+
+Note that a pattern of the form `Optional(T).None` is an expression pattern and
+is compared using `==`.
+
+### Templates
+
+If the type of the scrutinee of a pattern is either a template parameter or is
+an associated type that depends on a template parameter, any checking of the
+type of the scrutinee against the type of the pattern is deferred until the
+template parameter's value is known. During instantiation, patterns that are not
+meaningful due to a type error are instead treated as not matching. This
+includes cases where an `==` fails because of a missing `EqWith` implementation.
+
+```
+fn TypeName[template T:! Type](x: T) -> String {
+  match (x) {
+    // ✅ OK, the type of `x` is a template parameter.
+    case _: i32 => { return "int"; }
+    case _: bool => { return "bool"; }
+    case _: auto* => { return "pointer"; }
+    default => { return "unknown"; }
+  }
+}
+```
+
+```
+fn NeverWorks[template T:! Type](triple: (T, T, T)) {
+  match (triple) {
+    // ❌ Error: type mismatch matching struct against (T, T, T)
```

Comment:
Would this error be diagnosed at definition time or instantiation?

Reply:
My intent was to diagnose at template definition time. But per the previous comment I guess maybe that doesn't fit the model well. Removed.

Reply:
I think we should be able to say that these differences are diagnosed at definition time.

Reply:
OK, I've added this back. As noted in the previous comment, I don't think the definition of dependent from #2200 is sufficient here, but I think it's sufficient in this design doc to describe the intent -- if it's clear in the template definition that the pattern can never match, then that's an error -- and omit the case analysis with the details.